### PR TITLE
Fix S2197 FP: False Positive when collection size methods or properties are used

### LINKED ISSUE #4748: Rule S2197: false positive when `Array.Length` or `Enumerable.Count` are used
### Description

Rule S2197 raises false positives when `Array.Length` or `Enumerable.Count` are used.

### Repro steps

See: https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnModulus.cs#L23-L27

### Expected behavior

No issue raised for this occurrences.

### Actual behavior

Issue is raised

### Related information

* C#/VB.NET Plugins version: 8.26

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #4748

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 1]
```diff
@@ -50,17 +59,16 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+                c.ReportIssue(Diagnostic.Create(Rule, equalsExpression.GetLocation(), constantValue < 0 ? "negative" : "positive"));
             }
         }
 
         private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
             && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
+            && !IsUnsigned(modulus, semanticModel)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:70]: pro tip: now that you do this check, you can filter early based on the AST if it contains `Length` or `Count` before querying the semantic model

also, you seem to do 2 semantic model checks for the same `expression` - is it possible to do the job with just one?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 2]
```diff
@@ -34,6 +35,14 @@ public sealed class EqualityOnModulus : SonarDiagnosticAnalyzer
         private const string DiagnosticId = "S2197";
         private const string MessageFormat = "The result of this modulus operation may not be {0}.";
 
+        private static readonly string[] MethodsOrPropertiesReturningAlwaysPositiveInt =
+        {
+            "System.Array.Length",
+            "System.Collections.Generic.IEnumerable<int>.Count<int>()",
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:41]: It should not raise for any IEnumerable, not just the one of integers.

Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:41]: Rules S1155 and S3981 check similar things and might contain some things that could help you..

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 3]
```diff
@@ -34,6 +35,14 @@ public sealed class EqualityOnModulus : SonarDiagnosticAnalyzer
...
+            "System.Collections.Generic.IEnumerable<int>.LongCount<int>()",
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:42]: Ignorable: I am thinking if we should cover `String.Length` and `List.Capacity` or not. It might be an overkill and in the same time it is an FP

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 1 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:70]: I wanted to keep them separate since they are logically checking 2 different things but I see the potential perf improvement.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 4]
```diff
@@ -72,5 +80,12 @@ private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semant
             return type.IsAny(KnownType.UnsignedIntegers)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:80]: Lets create a field containing both the UnsignedIntegers and the UintPtr and using that with .IsAny().

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 5]
```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
             && binary.IsKind(SyntaxKind.ModuloExpression);
 
-        private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semantic)
+        private static bool ExpressionIsAlwaysPositive(ExpressionSyntax expression, SemanticModel semantic)
         {
             var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            var isUint = type.IsAny(KnownType.UnsignedIntegers)
+                          || type.Is(KnownType.System_UIntPtr);
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:82]: I think it's clearer if you do an early return based on isUint

```suggestion
            if (type.IsAny(KnownType.UnsignedIntegers, KnownType.System_UIntPtr))
            {
              return true;
             }
```

and then below you won't need `!isUint` and `return isUint`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 6]
```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+
+            var leftExpression = ((BinaryExpressionSyntax)expression).Left;
+            if (!isUint && CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x)))
+            {
+                var symbol = semantic.GetSymbolInfo(((BinaryExpressionSyntax)expression).Left).Symbol;
+                return IsCollectionSizeMethodOrProperty(symbol);
+            }
+            return isUint;
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:90]: this can then be
```suggestion
            return CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x)))
              && semantic.GetSymbolInfo(((BinaryExpressionSyntax)expression).Left).Symbol is { } symbol
              && IsCollectionSizeMethodOrProperty(symbol);

```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 7]
```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
         }
+
+        private static bool IsCollectionSizeMethodOrProperty(ISymbol symbol) =>
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:93]: IMO `IsCollectionSize` is a better name

method or property is an implementation detail that can be seen easily

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 8]
```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+            (CountName.Equals(symbol.Name) || LongCountName.Equals(symbol.Name))
+            && symbol is IMethodSymbol methodSymbol
+            && methodSymbol.IsExtensionMethod
+            && methodSymbol.ReceiverType != null
+            && methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
+
+        private static bool IsLengthProperty(ISymbol symbol) =>
+            (LengthName.Equals(symbol.Name) || LongLengthName.Equals(symbol.Name))
+            && symbol is IPropertySymbol propertySymbol
+            && (propertySymbol.ContainingType.Is(KnownType.System_Array) || propertySymbol.ContainingType.Is(KnownType.System_String));
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:109]: can use the `IsAny()` extension method

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 9]
```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+
+        private static bool IsCollectionCountProperty(ISymbol symbol) =>
+            CountName.Equals(symbol.Name)
+            && symbol is IPropertySymbol propertySymbol
+            && (propertySymbol.ContainingType.DerivesOrImplements(KnownType.System_Collections_Generic_ICollection_T)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:114]: will `Implements()` suffice, given that you're only checking interfaces? what's the use case for derives? I cannot see it in the UTs

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 9 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:114]: I kept also `Derives()` for safety. I think I will keep only `Implements()` to keep it simple for now.
WDYT?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 10]
```diff
@@ -50,27 +61,61 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
             var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            if (type.IsAny(KnownType.UnsignedIntegers) || type.Is(KnownType.System_UIntPtr))
+            {
+                return true;
+            }
+
+            var leftExpression = ((BinaryExpressionSyntax)expression).Left;
+            return CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x))
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:87]: I realize that `leftExpression.ToString()` is executed multiple times although it's not necessary - you can extract that

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 11]
```diff
@@ -50,27 +61,61 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+                c.ReportIssue(Diagnostic.Create(Rule, equalsExpression.GetLocation(), constantValue < 0 ? "negative" : "positive"));
             }
         }
 
         private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
             && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
+            && !ExpressionIsAlwaysPositive(modulus, semanticModel);
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:72]: the check `expression.RemoveParentheses() is BinaryExpressionSyntax binary` is done in `IsModulus()`

and  `var leftExpression = ((BinaryExpressionSyntax)expression).Left;` is done inside `ExpressionIsAlwaysPositive`

this makes the code hard to read

you could just inline `IsModulus` and then change the signature of `ExpressionIsAlwaysPositive` to receive `BinaryExpressionSyntax`

like this you would also avoid having the surprise that `expression` is actually a parenthesized expression (you don't have a test but I would expect it would throw when doing the cast in `ExpressionIsAlwaysPositive()`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 9 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs]: Update: After doing a quick test `Derives` is not really needed at all.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 12]
```diff
@@ -50,27 +61,61 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
             expression.RemoveParentheses() is BinaryExpressionSyntax binary
             && binary.IsKind(SyntaxKind.ModuloExpression);
 
-        private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semantic)
+        private static bool ExpressionIsAlwaysPositive(ExpressionSyntax expression, SemanticModel semantic)
         {
             var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            if (type.IsAny(KnownType.UnsignedIntegers) || type.Is(KnownType.System_UIntPtr))
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:81]: `KnownType.UnsignedIntegers` is an `ImmutableArray`.
I should first put all the elements from `KnownType.UnsignedIntegers` + the single element `KnownType.System_UIntPtr`  in one `IEnumerable<KnownType>` and then use `IsAny`. 
I don't think it really worth it here. Except if I'm missing something of course.

[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:81]: ah ok sorry

downside of not doing the CR in the IDE

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 2 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs]: Note: Rule S1155 is currently in the process of being refactored (to support VB.NET as well) #5106

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 13]
```diff
@@ -34,7 +37,15 @@ public sealed class EqualityOnModulus : SonarDiagnosticAnalyzer
         private const string DiagnosticId = "S2197";
         private const string MessageFormat = "The result of this modulus operation may not be {0}.";
 
-        private static readonly CSharpExpressionNumericConverter ExpressionNumericConverter = new CSharpExpressionNumericConverter();
+        private const string CountName = nameof(Enumerable.Count);
+        private const string LongCountName = nameof(Enumerable.LongCount);
+        private const string LengthName = "Length";
```
Comment (Corniel) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:42]: nameof(String.Length)

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:42]: I intentionally set it as `"Length"` because it represents both the `Array.Length` and the `String.Length`.
Although `nameof(String.Length)` would give the same result, it would probably be misleading.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 11 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs]: your update matches what I had in mind

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnModulus.cs [hunk 14]
```diff
@@ -20,11 +20,26 @@ public void isOdd(int x)
...
+
+            var list = new List<int>();
+            y = list.Count % 2 == 1; // Compliant - Count property is > 0;
+            y = list.Capacity % 2 == 1; // Compliant - Capacityproperty is > 0;
+
+            var someString = "HelloImAWhale";
+            y = someString.Length % 2 == 1; // Compliant - Length property is > 0;
+
+            y = FakeCount() % 2 == 1; // Noncompliant
         }
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnModulus.cs:41]: have a test w/ parens?
`(enumerableLong.Count() % 2) == 1`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 9 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs]: you can add that test to the UT to document the expected behavior

**Review by andrei-epure-sonarsource — APPROVED** [review]
I left 2 comments to add more tests, LGTM

leaving @csaba-sagi-sonarsource to finish the review

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 13 above)
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:42]: Why didn't you use nameof(Array.Length); if yo used the same approach for all the other names?

Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:42]: Lets just put a comment there please that it stands for both.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs [hunk 15]
```diff
@@ -50,27 +61,59 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
...
+            || IsListCapacityProperty(symbol);
+
+        private static bool IsEnumerableCountMethod(ISymbol symbol) =>
+            (CountName.Equals(symbol.Name) || LongCountName.Equals(symbol.Name))
+            && symbol is IMethodSymbol methodSymbol
+            && methodSymbol.IsExtensionMethod
+            && methodSymbol.ReceiverType != null
+            && methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
+
+        private static bool IsLengthProperty(ISymbol symbol) =>
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs:103]: I would  group these three checks IsLengthProperty(symbol) IsCollectionCountProperty(symbol) and IsListCapacityProperty(symbol) into a different method and do the `is IPropertySymbol` check once and use the `PropertySymbol` as an argument for them.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Last round of changes. :)

**Comment by Corniel** [comment]
How this rule behave on types that define the % operator themselves? May be just ignore them.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs (hunk 15 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs]: NITPICK: I think you could have left the three methods to divide the different types with different properties like: 

```
        private static bool IsLengthProperty(IPropertySymbol propertySymbol) =>
            (LengthName.Equals(propertySymbol.Name) || LongLengthName.Equals(propertySymbol.Name))
            && propertySymbol.ContainingType.IsAny(KnownType.System_Array, KnownType.System_String);

        private static bool IsCollectionCountProperty(IPropertySymbol propertySymbol) =>
            CountName.Equals(propertySymbol.Name)
            && (propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_ICollection_T)
                || propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_IReadOnlyCollection_T));

        private static bool IsListCapacityProperty(IPropertySymbol propertySymbol) =>
            ListCapacityName.Equals(propertySymbol.Name)
            && propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_IList_T);
```

**Review by csaba-sagi-sonarsource — APPROVED** [review]
Please take a look at my last comment, however I approved the PR, so if you think it does not make sense you can skip applying it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnModulus.cs [hunk 16]
```diff
@@ -37,8 +37,9 @@ public void isOdd(int x)
 
             y = FakeCount() % 2 == 1; // Noncompliant
 
-            y = (enumerableLong.Count() % 2) == 1; // Compliant
-            y = 1 == (x % 2); // Noncompliant {{The result of this modulus operation may not be positive.}}
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnModulus.cs]: why did you remove this test with the literal on the left?
